bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/datalayer/ComponentDataImpl.java

```diff
@@ -15,101 +15,209 @@
+     * The current data layer supplier.
+     */
+    @NotNull
+    private final DataLayerSupplier dataLayerSupplier;
+
+    /**
+     * Construct the data layer model.
+     *
+     * @param supplier The data layer supplier.
+     */
+    public ComponentDataImpl(@NotNull final DataLayerSupplier supplier) {
+        this.dataLayerSupplier = supplier;
     }
 
     @Override
+    @NotNull
     public String getId() {
-        return component.getId();
+        return this.getDataLayerSupplier().getId().get();
     }
 
     @Override
-    public String getParentId() {
-        return null;
+    @Nullable
+    public String getType() {
+        return this.getDataLayerSupplier()
+            .getType()
+            .map(Supplier::get)
+            .orElse(null);
     }
 
     @Override
-    public String getType() {
-        return component.getDataLayerType();
+    @Nullable
+    public Date getLastModifiedDate() {
+        return this.getDataLayerSupplier()
+            .getLastModifiedDate()
+            .map(Supplier::get)
+            .orElse(null);
     }
 
     @Override
+    @Nullable
+    public String getParentId() {
+        return this.getDataLayerSupplier()
+            .getParentId()
+            .map(Supplier::get)
+            .orElse(null);
+    }
+
+    @Override
+    @Nullable
     public String getTitle() {
-        return component.getDataLayerTitle();
+        return this.getDataLayerSupplier()
+            .getTitle()
+            .map(Supplier::get)
+            .orElse(null);
     }
 
     @Override
+    @Nullable
     public String getDescription() {
-        return component.getDataLayerDescription();
+        return this.getDataLayerSupplier()
+            .getDescription()
+            .map(Supplier::get)
+            .orElse(null);
     }
 
     @Override
-    public Date getLastModifiedDate() {
-        ValueMap valueMap = resource.adaptTo(ValueMap.class);
-        Calendar lastModified = null;
+    @Nullable
+    public String getText() {
+        return this.getDataLayerSupplier()
+            .getText()
+            .map(Supplier::get)
+            .orElse(null);
+    }
 
-        if (valueMap != null) {
-            lastModified = valueMap.get(JcrConstants.JCR_LASTMODIFIED, Calendar.class);
+    @Override
+    @Nullable
+    public String getLinkUrl() {
+        return this.getDataLayerSupplier()
+            .getLinkUrl()
+            .map(Supplier::get)
+            .orElse(null);
+    }
 
-            if (lastModified == null) {
-                lastModified = valueMap.get(JcrConstants.JCR_CREATED, Calendar.class);
-            }
-        }
+    @Override
+    @Nullable
+    public String[] getShownItems() {
+        return this.getDataLayerSupplier()
+            .getShownItems()
+            .map(Supplier::get)
+            .orElse(null);
+    }
 
-        if (lastModified != null) {
-            return lastModified.getTime();
-        }
+    @Override
+    @Nullable
+    public AssetData getAssetData() {
```

Comment:
This methode is specific to an image component. I would re-introduce the removed ImageDataImpl class.

Reply:
Does the explanation [below](https://github.com/adobe/aem-core-wcm-components/pull/1115#discussion_r499713421) clear up this review item, or do you still think changes are required?